### PR TITLE
Create core game implementation

### PR DESCRIPTION
**Pull request by peddyjet**
I created the Die class, capable of rolling itself and storing its result in an encapsulated property.

The Game class holds three dice, able to roll them all simultaneously, summate them, output their results, and continuously roll dice.

The Program static class holds the main method for the program, instantiating the game class and making it roll dice continuously. The main method also instantiates the testing class for the program, running it after the game is completed. The class also holds an instance of the Random class as a singleton, to prevent over-instantiation errors.

The Testing class holds various tests for the game, ensuring that both the Game class and the Die class work as intended.

### CODE REVIEW

**Comment by inkpin** [comment]
CMP1903_A1_2324 Code Review
For: Lewis Page (28043287)
By: William Inkpin (28072184) 

**Summary**:

Lewis’s code is well structured and handles the objective of the assessment well. The die class achieves its purpose of assigning a random value to the variable newDieValue (correctly formatted) and returns a number between 1 and the variable SidesOnDie (holds value 6) + 1, resulting in newDieValue only containing a number between 1-6.

The game class consists of an array named Die[] which can take 3 different values. The dice is then rolled, and the values are assigned to the array respectively. A variable is declared named: diceRollSum in the ReportDiceRolls method, allowing the program to sum all the roll values. A for loop will then compute the sum and print the results of the dice rolls to the console. Lewis has then implemented logic allowing the user to roll again via an infinite loop, giving the user the opportunity to roll forever, or break from the loop via input. In case of the user inputting a capital Y to reroll, the char value is set to lowercase, preventing issues. 

The main class creates an instance of the game class, and then runs the RollDiceContinuously method from the game class. Once the user has broken from the infinite loop mentioned before, an output is displayed providing the user the sum of all rolls. Finally, the testing class is initialised, and its method is run.

The testing class rolls its own values, instead of comparing to the values that the user is presented with, one possible improvement would be to obtain the values from the game class and ensure that those values are between 1 – 6 (even though the logic in the dice class prevents any values out of range). In contrast, the sum is compared to the rolls the user is presented with, therefore there is no chance of error whilst the sum appears to be correct. 

**Criteria**:

Have the requirements been met: **Yes**
Is the code formatted using the Style Guidelines correctly: **Yes**
Is the code easy to read: **Yes, comments are also very descriptive** 
Are different errors handled correctly: **Yes**

**Review by BobbyPaley — COMMENTED** [review]
The code is very well written (structuraly and efficiently) while following the style guidelines, with pleanty of comments making it easy to read through and understand what each line of code is acheving. One problem with the code is that in the Game class, is that it takes and IEnumerable as an argument for the dice rolls but the method is private and the dice are already a parameter of the class in turn making it not reliant on encapsulation. Another thing is that there is no error checking in the program, so if something goes wrong it will keep happening.

in conclusion all of the requierments have been met and the program runs efficiently with no errors when testing.

**Review by inkpin — COMMENTED** [review]
CMP1903_A1_2324 Code Review
For: Lewis Page (28043287)
By: William Inkpin (28072184)

Summary:

Lewis’s code is well structured and handles the objective of the assessment well. The die class achieves its purpose of assigning a random value to the variable newDieValue (correctly formatted) and returns a number between 1 and the variable SidesOnDie (holds value 6) + 1, resulting in newDieValue only containing a number between 1-6.

The game class consists of an array named Die[] which can take 3 different values. The dice is then rolled, and the values are assigned to the array respectively. A variable is declared named: diceRollSum in the ReportDiceRolls method, allowing the program to sum all the roll values. A for loop will then compute the sum and print the results of the dice rolls to the console. Lewis has then implemented logic allowing the user to roll again via an infinite loop, giving the user the opportunity to roll forever, or break from the loop via input. In case of the user inputting a capital Y to reroll, the char value is set to lowercase, preventing issues.

The main class creates an instance of the game class, and then runs the RollDiceContinuously method from the game class. Once the user has broken from the infinite loop mentioned before, an output is displayed providing the user the sum of all rolls. Finally, the testing class is initialised, and its method is run.

The testing class rolls its own values, instead of comparing to the values that the user is presented with, one possible improvement would be to obtain the values from the game class and ensure that those values are between 1 – 6 (even though the logic in the dice class prevents any values out of range). In contrast, the sum is compared to the rolls the user is presented with, therefore there is no chance of error whilst the sum appears to be correct.

Criteria:

Have the requirements been met: Yes
Is the code formatted using the Style Guidelines correctly: Yes
Is the code easy to read: Yes, comments are also very descriptive
Are different errors handled correctly: Yes

**Comment by peddyjet** [comment]
Thank you for your feedback. I have taken your comments into account, fixing all the problems you raised with the code.
